pkg/controller/istiocontrolplane/istiocontrolplane_controller.go

```diff
@@ -165,3 +192,101 @@ func (r *ReconcileIstioControlPlane) Reconcile(request reconcile.Request) (recon
+	CreateFunc: func(_ event.CreateEvent) bool {
+		// no action
+		return false
+	},
+	GenericFunc: func(_ event.GenericEvent) bool {
+		// no action
+		return false
+	},
+	DeleteFunc: func(e event.DeleteEvent) bool {
+		object, err := meta.Accessor(e.Object)
+		log.Debugf("got delete event for %s.%s", object.GetName(), object.GetNamespace())
+		if err != nil {
+			return false
+		}
+		if object.GetLabels()[OwnerNameKey] != "" {
+			return true
+		}
+		return false
+	},
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		object, err := meta.Accessor(e.ObjectNew)
+		log.Debugf("got update event for %s.%s", object.GetName(), object.GetNamespace())
+		if err != nil {
+			return false
+		}
+		if object.GetLabels()[OwnerNameKey] != "" {
+			return true
+		}
+		return false
+	},
+}
+
+func (r *ReconcileIstioControlPlane) getOrCreateReconciler(icp *v1alpha2.IstioControlPlane) (*helmreconciler.HelmReconciler, error) {
+	key := reconcilersMapKey(icp)
+	var err error
+	var reconciler *helmreconciler.HelmReconciler
+	if reconciler, ok := reconcilers[key]; ok {
+		reconciler.SetNeedUpdateAndPrune(false)
+		oldInstance := reconciler.GetInstance()
+		reconciler.SetInstance(icp)
+		if reconciler.GetInstance().GetGeneration() != oldInstance.GetGeneration() {
+			//regenerate the reconciler
+			if reconciler, err = r.factory.New(icp, r.client); err == nil {
+				reconcilers[key] = reconciler
+			}
+		}
+		return reconciler, err
+	}
+	//not found - generate the reconciler
+	if reconciler, err = r.factory.New(icp, r.client); err == nil {
+		reconcilers[key] = reconciler
+	}
+	return reconciler, err
+}
+
+// Watch changes for Istio resources managed by the operator
+func watchIstioResources(c controller.Controller) error {
+	for _, t := range []runtime.Object{
+		&corev1.ServiceAccount{TypeMeta: metav1.TypeMeta{Kind: "ServiceAccount", APIVersion: "v1"}},
+		&rbacv1.Role{TypeMeta: metav1.TypeMeta{Kind: "ClusterRole", APIVersion: "v1"}},
+		&rbacv1.RoleBinding{TypeMeta: metav1.TypeMeta{Kind: "ClusterRoleBinding", APIVersion: "v1"}},
+		&rbacv1.ClusterRole{TypeMeta: metav1.TypeMeta{Kind: "ClusterRole", APIVersion: "v1"}},
+		&rbacv1.ClusterRoleBinding{TypeMeta: metav1.TypeMeta{Kind: "ClusterRoleBinding", APIVersion: "v1"}},
+		&corev1.ConfigMap{TypeMeta: metav1.TypeMeta{Kind: "ConfigMap", APIVersion: "v1"}},
+		&corev1.Service{TypeMeta: metav1.TypeMeta{Kind: "Service", APIVersion: "v1"}},
+		&appsv1.Deployment{TypeMeta: metav1.TypeMeta{Kind: "Deployment", APIVersion: "v1"}},
+		&appsv1.DaemonSet{TypeMeta: metav1.TypeMeta{Kind: "DaemonSet", APIVersion: "v1"}},
+		&autoscalingv2beta1.HorizontalPodAutoscaler{TypeMeta: metav1.TypeMeta{Kind: "HorizontalPodAutoscaler", APIVersion: "v2beta1"}},
+		&admissionregistrationv1beta1.MutatingWebhookConfiguration{TypeMeta: metav1.TypeMeta{Kind: "MutatingWebhookConfiguration", APIVersion: "v1beta1"}},
+		&v1beta1.Deployment{TypeMeta: metav1.TypeMeta{Kind: "Deployment", APIVersion: "v1beta1"}},
```

Comment:
Is it possible to reuse the resources defined at https://github.com/istio/operator/blob/master/pkg/controller/istiocontrolplane/pruningdetails.go#L44-L103 ?

Reply:
Fixed